### ISSUE TITLE
Phone blocked doesn't interrupt connection

### ISSUE THREAD

**Issue by Ypot** [celzero/rethink-app#1347]
Behavior: 
Phone blocked.
An app allowed to connect, keeps downloading, isn't blocked by the firewall.

Expected behavior: when phone is blocked, the download should be interrupted. 

Settings:
DNS+Firewall active
Firewall is set to work only when phone is blocked.
No DNS, no firewall bypassed.
The app is allowed to connect.
No special rules added.

**Comment by ignoramous** [comment]
> An app allowed to connect, keeps downloading, isn't blocked by the firewall.

Existing connections aren't dropped. We could do so as the code to accomplish this already exists (#932), but that'd be too disruptive.

We could let the user toggle `Block active connections as rules change` ON or OFF as per their preference.

**Comment by Ypot** [comment]
Hi

My suggestion is to have the option to set a delay. 

For example: after the phone is blocked, to set how many minutes till the firewall disrupts all the established connections.

**Comment by Ypot** [comment]
Sorry, I don't get what this option means: 

Block active connections as rules change

**Comment by smexyy** [comment]
We require a feature that immediately blocks all network traffic, similar to NetGuard or Pcapdroid, when we disable the app’s Wi-Fi or data. This is a significant issue. For instance, if I were to access a harmful website and start downloading a large file, I would want to terminate the connection instantly. However, the connection remains active.

**Comment by Ypot** [comment]
I understand. I like how netguard works.
I think it would be:


> * We could let the user toggle ON or OFF as per their preference:
>   
>   * Either: `Block active connections as rules change`

I like this option.
I can't imagine why users would change rules if they don't want to change immediately active connections. 

>   * Or: `Block active connections when device is locked`

I would add a configurable delay: Users would set after how many minutes the device is locked, connections are blocked.

> * Alternatively, we can provide a button to _reset_ / _finish_ ALL active connections.

That would be another feature, compatible with the two other.

> Which option do you think makes more sense?

I suppose all of them... Too much work?

**Comment by ignoramous** [comment]
> can't imagine why users would change rules if they don't want to change immediately active connections.

Yeah you're right. In fact, Rethink already implements this for IP and domain rules. Just not for universal rules:
- #932 

> Users would set how many minutes must pass after the device is locked, for the connections to be blocked.

Such things confuse all but power users. We've already got too many open feature requests asking us to simplify the UI.

> Too much work?

Not really. The code to achieve this is already in production. Just that UI knobs don't exist. 

I'm inclined to close *active* connections *automatically* without user input.

**Comment by Ypot** [comment]
> > Users would set how many minutes must pass after the device is locked, for the connections to be blocked.
> 
> Such things confuse all but power users. We've already got too many open feature requests asking us to simplify the UI.
> 

[...]

> 
> I'm inclined to close _active_ connections _automatically_ without user input.

If you want to close active connections when phone is locked, the delay could be done in a way that users are not confused.

For example, Netguard closes all connections when screen is off. No confusion for users. 
But I recently discovered that it has the option to delay the internet blocking after the screen is off, which I found very useful. In resume, it is a great option for power users, while noobs just get the default behavior.

**Comment by luckygitt** [comment]
I think it's a no-brainer - if you decide to block an app or make any changes to the settings, they should be implemented immediately. Nothing else needed, no delays, no restart clicks, just do it.

**Comment by ignoramous** [comment]
> if you decide to block an app or make any changes to the settings, they should be implemented immediately

That already happens today. This bug is about terminating existing active connections because the device was *locked*.

**Comment by ignoramous** [comment]
> we need to terminate the active connection when toogle off any app wifi/data 

This *already* is the case today. If not, that's a bug.

**Comment by ignoramous** [comment]
Dropping connections once any (or at least *most*) of the rules apply will be part of `v055o`. We've completed all of the network engine side changes, and only a few Android (Kotlin) side changes (closing conns tied to an app/proxy on device events, rule change events etc) remain.

**Comment by ignoramous** [comment]
@hussainmohd-a close this bug if these features have been implemented in `v055o`+:
- Terminate active connections when device is locked.
- Terminate active connections to those IPs/domains for the app when per-app block domain/IP rules are added.
- Terminate active connections of the app when the app is set to (wifi/mobile) blocked / isolated.

**Comment by lylid** [comment]
Looking at this on v0.5.6, I think the situation is a bit different from
"not implemented yet".

`TunFlowManager.closeTrackedConnsOnDeviceLock()` already does exactly what this issue asks for. But its only call site is inside
`BraveVPNService.onUpstreamAnswer()`:

```kotlin
onDeviceLocked = { TunFlowManager.closeTrackedConnsOnDeviceLock(buildFlowContext()) },
```

That is the DNS answer path. So tracked connections are closed on lock only if a DNS answer happens to arrive while the device is locked.

The connections this issue is about are precisely the ones that never trigger it: once a websocket is established it issues no further DNS queries, so on an idle locked device the callback never fires. The function ends up being dead code for the exact case it was written for.

**Reproduce**

1. Enable `Block when device is locked`
2. Use an app that keeps a persistent connection (Signal is the obvious one)
3. Lock the device and leave it idle
4. Messages still arrive; the connection is still listed as active

New connections are correctly blocked, so the rule looks like it is working until you check an app that never needs a new connection.

**Fix**

Binding the same function to `ACTION_SCREEN_OFF`, gated on the existing `getBlockWhenDeviceLocked()` preference, makes the rule behave as documented. I have this running locally on v0.5.6 (built `assembleFdroidFullRelease`, installed on a Pixel running GrapheneOS / Android 17): the socket dies at screen-off and the app resyncs at unlock.

It is 26 lines in one file. Happy to open a PR.

One question before I do: you noted above that closing active connections automatically could be too disruptive. The patch as I have it reuses the existing `Block when device is locked` preference, so it only affects users who have already opted into lock-based blocking. If you would rather have it separate - e.g. `Block active connections when device is locked` as you suggested - say so and I will put it behind its own preference instead.

**Idle drain**

Worth adding, since this is the secondary reason people enable the rule: unplugged at 100% with the device left untouched overnight, the battery still read 100% after 8h - under 1% over the window, on a Pixel 10a with an active LTE connection and Signal installed.

That is a single observation rather than an A/B against an unpatched build, so treat it as indicative rather than measured. But it is consistent with the websocket actually going away instead of staying up all night.

**Comment by ignoramous** [comment]
> Binding the same function to ACTION_SCREEN_OFF, gated on the existing getBlockWhenDeviceLocked() preference, makes the rule behave as documented. I 

No, screen off is not a good signal of user absence (for instance, screen turns off due to proximity sensor while on voice calls).

Probably the `onDeviceLocked` flow needs to be decoupled from [`onUpstreanAnswer`](https://github.com/celzero/rethink-app/blob/21b8206b778f46ea2a5193f80b0414188ac78608/app/src/main/java/com/celzero/bravedns/tunnel/TunDnsManager.kt#L886) and [`evaluateFirewall`](https://github.com/celzero/rethink-app/blob/21b8206b778f46ea2a5193f80b0414188ac78608/app/src/main/java/com/celzero/bravedns/service/TunFlowManager.kt#L975). You're right that convoluted impl could be better, as noted in the comment above:

- Terminate active connections when device is locked.
- Terminate active connections to those IPs/domains for the app when per-app block domain/IP rules are added.
- Terminate active connections of the app when the app is set to (wifi/mobile) blocked / isolated.
